frame/header-mmr/src/lib.rs

```diff
@@ -62,83 +62,166 @@ mod tests;
-			// Update MMR and add mmr root to digest of block header
+					*progress = progress.saturating_add(*step);
+
+					Ok(T::DbWeight::get().reads_writes(1, *step * 2))
+				},
+			)
+			.map_or_else(|weight| weight, |weight| weight)
+		}
+
+		fn on_finalize(_: BlockNumberFor<T>) {
+			let parent_hash = <frame_system::Pallet<T>>::parent_hash();
+			let mut mmr = <Mmr<RuntimeStorage, T>>::new();
 			let _ = mmr.push(parent_hash);
 
-			if let Ok(parent_mmr_root) = mmr.get_root() {
-				if mmr.commit().is_ok() {
+			match mmr.finalize() {
+				Ok(parent_mmr_root) => {
 					let mmr_root_log = MerkleMountainRangeRootLog::<T::Hash> {
-						prefix: PARENT_MMR_ROOT_LOG_ID,
-						parent_mmr_root: parent_mmr_root.into(),
+						prefix: LOG_PREFIX,
+						parent_mmr_root,
 					};
 					let mmr_item = DigestItem::Other(mmr_root_log.encode());
 
 					<frame_system::Pallet<T>>::deposit_log(mmr_item.into());
-				} else {
-					log::error!("Commit MMR - FAILED");
 				}
-			} else {
-				log::error!("Calculate MMR - FAILED");
+				Err(e) => {
+					log::error!("Failed to finalize MMR due to {}", e);
+				}
 			}
 		}
 	}
 	#[pallet::call]
-	impl<T: Config> Pallet<T> {}
 	impl<T: Config> Pallet<T> {
+		#[pallet::weight(T::DbWeight::get().writes(1))]
+		pub fn config_pruning(
+			origin: OriginFor<T>,
+			step: Option<NodeIndex>,
+			progress: Option<NodeIndex>,
+			last_position: Option<NodeIndex>,
+		) -> DispatchResultWithPostInfo {
+			ensure_root(origin)?;
+
+			<PruningConfiguration<T>>::try_mutate(|c| {
+				let mut modified = false;
+
+				if let Some(step) = step {
+					c.step = step;
+					modified = true;
+				}
+				if let Some(progress) = progress {
+					c.progress = progress;
+					modified = true;
+				}
+				if let Some(last_position) = last_position {
+					c.last_position = last_position;
+					modified = true;
+				}
+
+				if modified {
+					Ok(().into())
+				} else {
+					Err("No changes".into())
+				}
+			})
+		}
+	}
+	impl<T: Config> Pallet<T> {
+		pub fn offchain_key(position: NodeIndex) -> Vec<u8> {
+			(T::INDEXING_PREFIX, position).encode()
+		}
+
 		darwinia_support::impl_rpc! {
 			pub fn gen_proof_rpc(
-				block_number_of_member_leaf: u64,
-				block_number_of_last_leaf: u64,
+				block_number_of_member_leaf: NodeIndex,
+				block_number_of_last_leaf: NodeIndex,
 			) -> RuntimeDispatchInfo<T::Hash> {
 				if block_number_of_member_leaf <= block_number_of_last_leaf {
-					let store = <ModuleMMRStore<T>>::default();
 					let mmr_size = mmr::leaf_index_to_mmr_size(block_number_of_last_leaf);
-					if mmr_size <= <MMRCounter<T>>::get() {
-						let mmr = <MMR<_, MMRMerge<T>, _>>::new(mmr_size, store);
-						let pos = mmr::leaf_index_to_pos(block_number_of_member_leaf);
 
-						if let Ok(merkle_proof) = mmr.gen_proof(vec![pos]) {
+					if mmr_size <= <MmrSize<T>>::get() {
+						let position = mmr::leaf_index_to_pos(block_number_of_member_leaf);
+						let mmr = <Mmr<OffchainStorage, T>>::with_size(<MmrSize<T>>::get());
```

Comment:
Not right using `<MmrSize<T>>::get()` here?

`<Mmr<OffchainStorage, T>>::with_size(<MmrSize<T>>::get()).gen_proof(position)` is different with

`<Mmr<OffchainStorage, T>>::with_size(mmr_size).gen_proof(position)`

mmr_size here is `mmr::leaf_index_to_mmr_size(block_number_of_last_leaf)`

Reply:
Good catch. That's why I found `gen_proof` slower than before.